python.d/traefik.chart.py

```diff
@@ -0,0 +1,196 @@
+        'lines': [
+            ['uptime_sec', None, 'absolute']
+        ]}
+    }
+
+HEALTH_STATS = [
+    'uptime_sec',
+    'average_response_time_sec',
+    'total_response_time_sec',
+    'total_count',
+    'total_status_code_count'
+]
+
+class Service(UrlService):
+    def __init__(self, configuration=None, name=None):
+        UrlService.__init__(self, configuration=configuration, name=name)
+        self.url = self.configuration.get('url', 'http://localhost:8080/health')
+        self.order = ORDER
+        self.definitions = CHARTS
+        self.data = {
+            'None': 0,
+            'average_response_time_per_iteration_sec': 0
+        }
+
+    def check(self):
+        if not (self.url):
+            self.error('url is not defined in the module configuration file')
+            return False
+
+        self._manager = self._build_manager()
+        data = self._get_raw_data(self.url)
+        data = loads(data)
+
+        for code in data['total_status_code_count']:
+            self.definitions['detailed_response_codes']['lines'].append(
+                    [code, code, 'incremental']
+            )
+            if code not in self.data:
+                self.data[code] = data['total_status_code_count'][code]
+        return UrlService.check(self)
+
+    def _get_data(self):
+        data = self._get_raw_data()
+
+        if not data:
+            return None
+
+        data = loads(data)
+
+        self.get_data_per_code_status(raw_data=data)
+
+        self.get_data_per_code_family(raw_data=data)
+
+        self.get_data_per_code(raw_data=data)
+
+        to_netdata = fetch_data_(raw_data=data, metrics=HEALTH_STATS)
+        to_netdata.update(self.data)
+
+        self.data['average_response_time_per_iteration_sec'] = (data['total_response_time_sec'] * 1000000) / data['total_count']
+        to_netdata.update(self.data)
+
+        return to_netdata or None
+
+    def get_data_per_code_status(self, raw_data):
+        code_status = {'successful_requests': 0, 'redirects': 0,
+                       'bad_requests': 0, 'server_errors': 0,
+                       'other_requests': 0}
+        for code in raw_data['total_status_code_count']:
+            code_prefix = code[0]
+            if code_prefix == '1' or code_prefix == '2' or code == '304':
+                code_status['successful_requests'] += raw_data['total_status_code_count'][code]
+            elif code_prefix == '3':
+                code_status['redirects'] += raw_data['total_status_code_count'][code]
+            elif code_prefix == '4':
+                code_status['bad_requests'] += raw_data['total_status_code_count'][code]
+            elif code_prefix == '5':
+                code_status['server_errors'] += raw_data['total_status_code_count'][code]
+            else:
+                code_status['other_requests'] += raw_data['total_status_code_count'][code]
+        self.data.update(code_status)
```

Comment:
since we have all the stuff in `self.data` this method should be:

```python
from collections import defaultdict
...
...
...
def get_data_per_code_status(self, raw_data):
    data = defaultdict(int)
    for code, value in raw_data['total_status_code_count'].items():
        code_prefix = code[0]
        if code_prefix == '1' or code_prefix == '2' or code == '304':
            data['successful_requests'] += value
        ...
        ...
    self.data.update(data)
```

```